torch/nn/functional.py

```diff
@@ -3839,27 +3839,133 @@ def fold(input, output_size, kernel_size, dilation=1, padding=0, stride=1):
+    end are used to pad the beginning. The first and second axes of the tensor
+    are not padded.
+
+    Args:
+        input: Tensor with shape :math:`(N, C, D[, H, W])`.
+        padding: Tuple containing the number of elements to pad each side of
+            the tensor. The length of padding must be twice the number of
+            paddable axes. For example, the length of padding should be 4 for a
+            tensor of shape :math:`(N, C, H, W)`, and the length should be 6
+            for a tensor of shape :math:`(N, C, D, H, W)`.
+
+    Examples::
+
+        >>> x = torch.arange(6).view(1, 1, 2, 3)  # Create tensor
+        >>> # Example 1
+        >>> padding = (1, 1, 1, 1)
+        >>> y = F.pad(x, padding, mode='circular')
+        >>> print(x)
+        >>> print(y)
+        >>> print(y.shape)  # torch.Size([1, 1, 4, 5])
+        >>> # Example 2
+        >>> padding = (1, 1, 2, 2)
+        >>> z = F.pad(x, padding, mode='circular')
+        >>> print(z)
+        >>> print(z.shape)  # torch.Size([1, 1, 6, 5])
     """
-    Arguments
-        :param input: tensor of shape :math:`(N, C_{\text{in}}, H, [W, D]))`
-        :param padding: (tuple): m-elem tuple where m is the degree of convolution
-    Returns
-        :return: tensor of shape :math:`(N, C_{\text{in}}, [D + 2 * padding[0],
-                 H + 2 * padding[1]], W + 2 * padding[2]))`
-    """
+    shape = input.shape
+    paddable_shape = shape[2:]
+    ndim = len(paddable_shape)
 
-    input = torch.cat([input, input[:, :, 0:padding[-1]]], dim=2)
-    input = torch.cat([input[:, :, -(padding[-1] + padding[-2]):-padding[-1]], input], dim=2)
+    for idx, size in enumerate(paddable_shape):
+        # Only supports wrapping around once
+        assert padding[-(idx * 2 + 1)] <= size
+        assert padding[-(idx * 2 + 2)] <= size
+        # Negative padding should not result in negative sizes
+        assert (padding[-(idx * 2 + 1)] + padding[-(idx * 2 + 2)] + size) >= 0
 
-    if len(padding) > 2:
-        input = torch.cat([input, input[:, :, :, 0:padding[-3]]], dim=3)
-        input = torch.cat([input[:, :, :, -(padding[-3] + padding[-4]):-padding[-3]], input], dim=3)
+    # Get shape of padded array
+    padded_shape = shape[:2]
+    for idx, size in enumerate(paddable_shape):
+        padded_shape += (size + padding[-(idx * 2 + 1)] + padding[-(idx * 2 + 2)],)
 
-    if len(padding) > 4:
-        input = torch.cat([input, input[:, :, :, :, 0:padding[-5]]], dim=4)
-        input = torch.cat([input[:, :, :, :, -(padding[-5] + padding[-6]):-padding[-5]], input], dim=4)
+    out = torch.empty(padded_shape, dtype=input.dtype, layout=input.layout,
+                      device=input.device)
+
+    # Put original array in padded array
+    if ndim == 1:
+        out_d0 = max(padding[-2], 0)
+        out_d1 = padded_shape[2] - max(padding[-1], 0)
+
+        in_d0 = max(-padding[-2], 0)
+        in_d1 = shape[2] - max(-padding[-1], 0)
```

Comment:
It looks like this section is common to all branches. Can it be pulled out of the if/elif block?

Reply:
Actually I guess this is fine. We don't need to code golf this. It's very readable as is.

Reply:
I made a local function that does these index calculations. I still have the same if block but the code is less redundant now.